File: streaming-kafka-course/src/main/scala/com/backwards/kafka/streaming/ProducerDemo.scala
package com.backwards.kafka.streaming

import java.util.Properties
import org.apache.kafka.clients.producer.ProducerConfig.CLIENT_ID_CONFIG
import org.apache.kafka.clients.producer.{KafkaProducer, ProducerRecord}
import com.backwards.kafka.streaming.Config._
import com.typesafe.scalalogging.LazyLogging

object ProducerDemo extends App with LazyLogging {
  // TODO - Extract to scala-backwards

  implicit val map2Properties: Map[String, String] => Properties =
    m => (new Properties /: m) {
    case (properties, (k, v)) =>
      properties.put(k, v)
      properties
  }

  val kafkaProps = load[Map[String, String]]("kafka") + (CLIENT_ID_CONFIG -> "producer-demo")

  val producer = new KafkaProducer[String, String](kafkaProps)

  val record = new ProducerRecord[String, String]("test-topic", "my-key", "my-value")

  producer.send(record)

  producer.close()
}

File: courses/beginners-course/src/main/java/com/backwards/kafka/ProducerDemo.java
package com.backwards.kafka;

import java.net.URI;
import java.net.URISyntaxException;
import scala.collection.immutable.Seq;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.common.serialization.StringSerializer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import com.backwards.config.BootstrapConfig;
import io.lemonlabs.uri.Uri;
import io.lemonlabs.uri.Uri$;
import static java.util.Collections.singletonList;
import static org.apache.kafka.clients.producer.ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG;
import static org.apache.kafka.clients.producer.ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG;
import static scala.jdk.CollectionConverters.ListHasAsScala;

public class ProducerDemo {
    private static final Logger logger = LoggerFactory.getLogger(ProducerDemo.class);

    private static final String topic = "first_topic";

    public static void main(String[] args) throws URISyntaxException {
        produce(config());
    }

    private static KafkaConfig config() throws URISyntaxException {
        Seq<Uri> bootstrapServers = ListHasAsScala(singletonList(Uri$.MODULE$.apply(new URI("http://127.0.0.1:9092")))).asScala().toSeq();

        return KafkaConfig.apply(new BootstrapConfig(bootstrapServers))
                .add(KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName())
                .add(VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
    }

    private static void produce(KafkaConfig config) {
        KafkaProducer<String, String> producer = new KafkaProducer<>(config.toProperties());

        for (int i = 0; i < 10; i++) {
            String key = "id_" + i;
            String value = "hello world " + i;

            final ProducerRecord<String, String> record = new ProducerRecord<>(topic, key, value);

            // The following is asynchronous - if we do not flush or flush and close, then this app can end before data is actually sent.
            producer.send(record, (recordMetadata, e) -> {
                if (e == null) {
                    logger.info(
                        "\nReceived Metadata for '" + record.key() + "'\n" +
                        "- Topic: " + recordMetadata.topic() + "\n" +
                        "- Partition: " + recordMetadata.partition() + "\n" +
                        "- Offset: " + recordMetadata.offset() + "\n" +
                        "- Timestamp: " + recordMetadata.timestamp()
                    );
                } else {
                    logger.error("Error while producing", e);
                }
            });
        }

        // Either flush:
        // producer.flush();
        // Or flush and close:
        producer.close();
    }
}
